SwiftPM Snippets tutorial (#2)
swift-on-server/articles: Sources/Articles/Documentation.docc/2024/swiftpm-snippets/SwiftPM Snippets.md

# Getting started with SwiftPM Snippets

> [SwiftPM Snippets](https://github.com/apple/swift-evolution/blob/main/proposals/0356-swift-snippets.md) are one of the most powerful features of the [Swift Package Manager](https://github.com/apple/swift-package-manager), and yet two years after their introduction few developers know they exist. This tutorial will explain some of the advantages of using SwiftPM Snippets and show you how to add Snippets to a Swift package, preview them in documentation locally using [DocC](https://github.com/apple/swift-docc), render them in [Unidoc](https://github.com/tayloraswift/swift-unidoc), and publish them on your own website or a centralized platform like [Swiftinit](https://swiftinit.org) where readers can view their code and interact with the symbols contained within them.

Reviewer: Also the article is mostly about DocC + Snippet previews, Unidoc examples are currently missing... feels like we should create a follow-up article focusing on Unidoc.

> [SwiftPM Snippets](https://github.com/apple/swift-evolution/blob/main/proposals/0356-swift-snippets.md) are one of the most powerful features of the [Swift Package Manager](https://github.com/apple/swift-package-manager), and yet two years after their introduction few developers know they exist. This tutorial will explain some of the advantages of using SwiftPM Snippets and show you how to add Snippets to a Swift package, preview them in documentation locally using [DocC](https://github.com/apple/swift-docc), render them in [Unidoc](https://github.com/tayloraswift/swift-unidoc), and publish them on your own website or a centralized platform like [Swiftinit](https://swiftinit.org) where readers can view their code and interact with the symbols contained within them.

Author: done

## What are Swift Snippets?

Swift Snippets were invented in 2022 by [Ashley Garland](https://github.com/bitjammer) and first shipped in Swift 5.7. [Originally conceived](https://forums.swift.org/t/se-0356-swift-snippets/57097) as a way to validate example programs by compiling them, developers have since found a variety of additional applications for them, ranging from testing to documentation to even full-blown prototyping of multi-module setups.

Despite their immense potential, very little documentation exists on how to use Swift Snippets, and awareness of the feature remains [surprisingly low](https://github.com/search?q=%40Snippet+language%3Aswift&type=code).

> The outlook for this feature is pessimistic considering the low adoption rate. This might change if Apple invests in better documentation and adopts Snippets in their own repositories.
>
> — [A critical look at Swift Snippets](https://blog.eidinger.info/a-critical-look-at-swift-snippets-swift-57) by Marco Eidinger

> 

Author: done

    Some documentation engines such as Unidoc can render Snippets in the browser with linked identifiers, allowing readers to interact with the symbols in the code.

## Adding Snippets to a Swift package

For this tutorial, we will create a package named `snippets-example`.

```bash
$ mkdir snippets-example
$ cd snippets-example
$ swift package init --name 'Swift Snippets'
```

This should initialize a new Swift package with a `Package.swift` resembling the following.

@Code(name: "Package.swift", file: Manifest.1.swift)

Rename the library target to `SnippetsExample`.

@Code(name: "Package.swift", file: Manifest.2.swift)

Next, create a new directory named `Snippets` at the top level of the package.

```bash
$ mkdir Snippets
```

>   Note: Despite the specification of [SE-0356](https://github.com/apple/swift-evolution/blob/main/proposals/0356-swift-snippets.md#overriding-the-location-of-snippets), it is not possible to name the directory anything other than `Snippets`.

Inside the `Snippets` directory, create a new Swift file named `SnippetsExample_I.swift`.

@Snippet(id: "SnippetsExample_I")

> You should now have a directory structure that looks like this:

Author: see Slack discussion about active voice

```
📂 swift-snippets
├── 📂 Snippets
│   └── 📄 SnippetsExample_I.swift
├── 📂 Sources
│   └── 📂 SnippetsExample
│       └── 📄 anchor.swift
├── 📄 Package.swift
└── 📄 .gitignore
```

That’s it!

## Running Snippets

The `swift build` command will automatically discover and compile all Snippets in a package.

Snippets are just modules with discovery enabled, which means you can also build a single Snippet with the `--target` flag.

```bash
$ swift build --target SnippetsExample_I
```

> You can run a Snippet with the `swift run` command.

Author: see Slack discussion about active voice

```bash
$ swift run SnippetsExample_I
```

```text
Building for debugging...
[5/5] Linking SnippetsExample_I
Build complete! (0.97s)
Hi Barbie!
```

> You could also run a Snippet with release optimizations, just like any other executable target.

Reviewer: ```suggestion
A snippet can also be run with release optimizations, just like any other executable target.
```

Author: ditto

```bash
$ swift run -c release SnippetsExample_I
```

```text
Building for production...
[17/17] Linking SnippetsExample_I
Build complete! (9.73s)
Hi Barbie!
```

>   Note:
>   If the Snippet name contains special characters, you should pass the file name as-is to the `--target` flag. For example, to run a Snippet named `Foo bar.swift`, you would use `swift run 'Foo bar'` and not `Foo_bar`.

## Embedding Snippets in documentation

Most modern documentation engines support embedding Snippets in Markdown documentation via the `@Snippet` block directive.

Let’s create a documentation bundle for the `SnippetsExample` target.

```bash
$ mkdir -p Sources/SnippetsExample/docs.docc
```

>   Tip:
>   You can name the documentation bundle anything you like, as long as it has a `.docc` extension. The name comes from the DocC tool, but most other documentation engines will recognize it too.

Create a markdown article named `My article.md` in the `docs.docc` directory.

> @Code(name: "My article.md", file: "My article (2).md.txt")

Reviewer: Shouldn't we use the articles in sequential order, 1,2,3 instead of 2,1,6? 🤔

Author: done

> In the example above, we have specified the Snippet to include by `path` identity. Despite its naming, the `path` syntax is not a file path. The first component is the name of the package **as specified by the ``PackageDescription/Package/name`` field in the manifest**. The second component is always the string `Snippets`. The third component is the Snippet ID, which is the name of the Snippet file without the `.swift` extension. If the Snippet ID contains special characters, you should pass the ID as-is, without replacing any characters.

Reviewer: ```suggestion
In the example above, the Snippet is included by using the `path` argument. 

Despite its naming, the `path` syntax is not a file path. The first component is the name of the package **as specified by the ``PackageDescription/Package/name`` field in the manifest**. The second component is always the string `Snippets`. The third component is the Snippet ID, which is the name of the Snippet file without the `.swift` extension.

If the Snippet ID contains special characters, the ID should be passed as-is, without replacing any characters.
```

Author: broke up the paragraph, but kept active voice

Some documentation engines such as Unidoc support referencing Snippets by `id`.

@Code(name: "My article.md", file: "My article (1).md.txt")

>   Important:
>   DocC does not currently support referencing Snippets by ID. Instead, you must use the fully-qualified `path` syntax to reference a Snippet.

> You should now have a project layout that looks like this:

Reviewer: ```suggestion
The project layout should now look like this:
```

Author: done

```
📂 swift-snippets
├── 📂 Snippets
│   └── 📄 SnippetsExample_I.swift
├── 📂 Sources
│   └── 📂 SnippetsExample
│       ├── 📂 docs.docc
│       │   └── 📄 My article.md
│       └── 📄 anchor.swift
├── 📄 Package.swift
└── 📄 .gitignore
```

### Previewing Snippets with DocC

> Many developers find [DocC](https://github.com/apple/swift-docc) helpful for previewing documentation locally. To use DocC, add the [swift-docc-plugin](https://github.com/apple/swift-docc-plugin) to the package manifest.

Reviewer: We should mention here (again), that it is possible to build the DocC documentation using Xcode, but currently snippets won't render, due to the bug.

Author: done

@Code(name: Package.swift, file: Manifest.3.swift)

> You can then launch DocC with the `preview-documentation` package subcommand:

Reviewer: The documentation preview can be launched using the `preview-documentation` subcommand.

We should also mention why we need the --disable-sandbox flag.

Also we have to create an empty .swift file under the SnippetsExample target, otherwise the package won't compile, since SPM can't find any sources.

Author: added a note about empty file earlier in the tutorial, but i personally have no idea why `--disable-sandbox` is needed, i have cargo culting that for a long time. the [official DocC README](https://github.com/swiftlang/swift-docc-plugin?tab=readme-ov-file#previewing-documentation) doesn’t mention why it is needed.

```bash
$ swift package --disable-sandbox preview-documentation --target SnippetsExample
```

> You can find the rendered article at [`http://localhost:8080/documentation/snippetsexample/my-article`](http://localhost:8080/documentation/snippetsexample/my-article).

Reviewer: ```suggestion
The rendered article should be available at [`http://localhost:8080/documentation/snippetsexample/my-article`](http://localhost:8080/documentation/snippetsexample/my-article).
```

Author: ditto

>   Warning:
>   At the time of writing, there is a [known bug](https://github.com/apple/swift-docc/issues/944) causing DocC preview to crash if the target name contains special characters.

## Using Snippet captions

If a Snippet begins with contiguous line comments, those comments will be parsed as Markdown and treated as a Snippet caption. Try adding the following code to a Snippet named `SnippetsExample_II.swift`.

@Code(name: "SnippetsExample_II.swift", file: SnippetsExample_II.swift)

When embedded, it should look like this:

---

@Snippet(id: SnippetsExample_II)

>   Note:
>   Snippet captions cannot embed other Snippets.

## Redacting parts of a Snippet

> You can redact portions of a Snippet using **slice directives**. A slice directive is a line comment token that starts with `snippet` followed by a dot and an identifier. The following identifiers have special meanings:

Reviewer: ```suggestion
It is possible to redact portions of a Snippet using **slice directives**. A slice directive is a line comment token that starts with `snippet` followed by a dot and an identifier. The following identifiers have special meanings:
```

| Identifier        | Behavior                                   |
| ----------------- | ------------------------------------------ |
| `snippet.hide`    | Hides the content following the directive. |

> | `snippet.end`     | Hides the content following the directive. |

Reviewer: ```suggestion
| `snippet.end`     | Ends the current slice. |
```

Author: this was repeated intentionally, from an implementation standpoint `snippet.end` and `snippet.hide` are identical.

Reviewer: Ah ok, you are completely right about this. :)

| `snippet.show`    | Shows the content following the directive. |

Below is an example of a Snippet that uses redactions to hide the `import` statements.

@Code(name: SnippetsExample_III.swift, file: SnippetsExample_III.swift)

When embedded, it should look like this:

---

@Snippet(id: SnippetsExample_III)

>   Warning:
>   At the time of writing, there is a [known bug](https://github.com/apple/swift-docc/issues/946) preventing DocC from slicing Snippets that contain multi-line string literals correctly.

### Slice indentation

The indentation of the first `snippet.show` determines specifies the maximum amount of indentation to remove from the Snippet. In the example below, four spaces of indentation will be removed from the rendered Snippet. Note that the `snippet.end` token is required in order to prevent the Snippet from including the final brace, which would have prevented the indentation from being removed.

@Code(name: SnippetsExample_IV.swift, file: SnippetsExample_IV.swift)

When embedded, it should look like this:

---

@Snippet(id: SnippetsExample_IV)

## Using named slices

> You can also use **named slices** to create Snippets with multiple embeddable sections. It’s a good idea to give slices uppercase names to distinguish them from special slice directives.

Reviewer: ```suggestion
Named slices can also be used to create Snippets with multiple embeddable sections. It’s a good idea to give slices uppercase names to distinguish them from special slice directives.
```

Below is an example of a Snippet with a caption and three named slices.

@Code(name: SnippetsExample_V.swift, file: SnippetsExample_V.swift)

> Here’s how you might embed the slices in a Markdown article.

Reviewer: ```suggestion
Here’s how the slices can be embedded in a Markdown article.
```

@Code(name: "My article.md", file: "My article (6).md.txt")

And here’s how the embedded slices should look.

---

@Snippet(id: SnippetsExample_V, slice: DECLARATION)
@Snippet(id: SnippetsExample_V, slice: BODY)
@Snippet(id: SnippetsExample_V, slice: EXIT)

## Rendering Snippets with linkable references

> Some documentation engines such as Unidoc support rendering Snippets with linked identifiers. This is possible because Snippets are compiled along with the rest of the package, allowing them to be mapped by [IndexStoreDB](https://github.com/apple/indexstore-db).

Reviewer: It's a bit strange, that here we start mentioning Unidoc again, but so far only DocC examples were provided, why don't we create a standalone follow-up article about using Unidoc & snippets and keep this one simple (Snippets + DocC). 🤔

Author: i rewrote the entire section as a **Where to go from here** and removed references to IndexStoreDB

> Some documentation engines such as Unidoc support rendering Snippets with linked identifiers. This is possible because Snippets are compiled along with the rest of the package, allowing them to be mapped by [IndexStoreDB](https://github.com/apple/indexstore-db).

Reviewer: What's IndexStoreDB & the SwiftInit platform? As a newcomer, I don't necessary have the knowledge and I don't know about these things at all, it'd be nice if you could provide a short 1 sentence explanation, or create a standalone article about Unidoc, IndexStoreDB + Swiftinit.

Author: see above
